File: legacy/ciphers/nihilist.py
#plaintext: string
#polybius_keyword: string
#keyword: string
def encode_nilist(plaintext, polybius_keyword, keyword):
    alphabet = "ABCDEFGHIKLMNOPQRSTUVWXYZ"
    
    plaintext = plaintext.upper()
    plaintext = [i for i in plaintext if i in alphabet]
    polybius_keyword = polybius_keyword.upper()
    keyword = keyword.upper()
    
    val_map = {}
    index = 1
    for letter in polybius_keyword:
       if letter not in val_map and letter != 'J':
           val_map[letter] = index
           index += 1
           
    for letter in alphabet:
        if letter not in val_map:
            val_map[letter] = index
            index += 1
    
    result = []
    for i in range(len(plaintext)):
        if plaintext[i] not in alphabet:
            continue
        in1 = val_map[plaintext[i]]
        in2 = val_map[keyword[i % len(keyword)]]
        result.append(10 * ((in1 + 4) // 5) + (in1 - 1) % 5 + 1 +
                      10 * ((in2 + 4) // 5) + (in2 - 1) % 5 + 1)
    
    return result

#ciphertext: array of ints
#polybius_keyword: string
#keyword: string
def decode_nilist(ciphertext, polybius_keyword, keyword):
    alphabet = "ABCDEFGHIKLMNOPQRSTUVWXYZ"
    
    polybius_keyword = polybius_keyword.upper()
    keyword = keyword.upper()
    
    val_map = {}
    char_map = {}
    index = 1
    for letter in polybius_keyword:
       if letter not in val_map and letter != 'J':
           val_map[letter] = index
           char_map[10 * ((index + 4) // 5) + (index - 1) % 5 + 1] = letter
           index += 1
           
    for letter in alphabet:
        if letter not in val_map:
            val_map[letter] = index
            char_map[10 * ((index + 4) // 5) + (index - 1) % 5 + 1] = letter
            index += 1
    
    result = []
    for i in range(len(ciphertext)):
        in1 = ciphertext[i]
        in2 = val_map[keyword[i % len(keyword)]]
        result.append(in1 - 10 * ((in2 + 4) // 5) - (in2 - 1) % 5 - 1)
    
    result = [char_map[index] for index in result]
    
    return result
    
   

print(encode_nilist("My wife dresses to kill. She cooks the same way.", "youngman", "HENNY"))
       
print(decode_nilist(encode_nilist("My wife dresses to kill. She cooks the same way.", "youngman", "HENNY") , "youngman", "HENNY"))

File: legacy/public/js/ciphers/nihilist.js
export function encodeNihilist(plaintext, polybiusKeyword, keyword) {
    const alphabet = "ABCDEFGHIKLMNOPQRSTUVWXYZ";
    
    plaintext = plaintext.toUpperCase().split('').filter(i => alphabet.includes(i));
    polybiusKeyword = polybiusKeyword.toUpperCase();
    keyword = keyword.toUpperCase();
    
    const valMap = {};
    let index = 1;
    for (let letter of polybiusKeyword) {
        if (!(letter in valMap) && letter !== 'J') {
            valMap[letter] = index;
            index++;
        }
    }
    
    for (let letter of alphabet) {
        if (!(letter in valMap)) {
            valMap[letter] = index;
            index++;
        }
    }
    
    const result = [];
    for (let i = 0; i < plaintext.length; i++) {
        if (!alphabet.includes(plaintext[i])) {
            continue;
        }
        const in1 = valMap[plaintext[i]];
        const in2 = valMap[keyword[i % keyword.length]];
        result.push(10 * Math.floor((in1 + 4) / 5) + (in1 - 1) % 5 + 1 +
                    10 * Math.floor((in2 + 4) / 5) + (in2 - 1) % 5 + 1);
    }
    
    return result;
}

export function decodeNihilist(ciphertext, polybiusKeyword, keyword) {
    const alphabet = "ABCDEFGHIKLMNOPQRSTUVWXYZ";
    
    polybiusKeyword = polybiusKeyword.toUpperCase();
    keyword = keyword.toUpperCase();
    
    const valMap = {};
    const charMap = {};
    let index = 1;
    for (let letter of polybiusKeyword) {
        if (!(letter in valMap) && letter !== 'J') {
            valMap[letter] = index;
            charMap[10 * Math.floor((index + 4) / 5) + (index - 1) % 5 + 1] = letter;
            index++;
        }
    }
    
    for (let letter of alphabet) {
        if (!(letter in valMap)) {
            valMap[letter] = index;
            charMap[10 * Math.floor((index + 4) / 5) + (index - 1) % 5 + 1] = letter;
            index++;
        }
    }
    
    const result = [];
    for (let i = 0; i < ciphertext.length; i++) {
        const in1 = ciphertext[i];
        const in2 = valMap[keyword[i % keyword.length]];
        result.push(in1 - 10 * Math.floor((in2 + 4) / 5) - (in2 - 1) % 5 - 1);
    }
    
    return result.map(index => charMap[index]);
}

export function nihilistSession(quote, polybiusKeyword, keyword) {
    const ciphertext = encodeNihilist(quote, polybiusKeyword, keyword);
    console.log(quote);

    const output = document.getElementById('check');
    output.innerHTML = 'Check';

    const grid = document.getElementById('grid');
    grid.innerHTML = ''; 

    const monosubstitution = document.getElementById('monosubstitution-output');
    monosubstitution.innerHTML = '';

    let chart = document.getElementById('frequency-output');
    chart.innerHTML = '';

    let leftOutput = document.getElementById('letters-left-output');
    leftOutput.innerHTML = '';

    const desc = document.getElementById('description');
    desc.innerHTML = `Solve this nihilist cipher with keyword ${keyword} and polybius keyword ${polybiusKeyword}`;

    ciphertext.forEach((letter) => {
        let div = document.createElement('div');
        div.className = 'grid-item';

        let inner_cipher_div = document.createElement('div');
        inner_cipher_div.className = 'cipher-letter';
        inner_cipher_div.innerText = letter;

        let inner_plain_div = document.createElement('div');
        inner_plain_div.className = 'plain-letter';
        inner_plain_div.tabIndex = '0';
        inner_plain_div.innerText = ' ';
        
        div.appendChild(inner_cipher_div);
        div.appendChild(inner_plain_div);
        grid.appendChild(div);
    });
}